src/objects/InstancedMesh.js

```diff
@@ -0,0 +1,374 @@
+/**
+ * @author pailhead / www.dusanbosnjak.com
+ */
+
+import { Mesh } from './Mesh';
+import { Object3D } from '../core/Object3D';
+import { InstancedBufferGeometry } from '../core/InstancedBufferGeometry';
+import { InstancedBufferAttribute } from '../core/InstancedBufferAttribute';
+import { Matrix3 } from '../math/Matrix3';
+import { Matrix4 } from '../math/Matrix4';
+import { Vector3 } from '../math/Vector3';
+import { Quaternion } from '../math/Quaternion';
+
+import { MeshDepthMaterial } from '../materials/MeshDepthMaterial';
+import { RGBADepthPacking } from '../constants';
+import { ShaderMaterial } from '../materials/ShaderMaterial';
+import { UniformsUtils } from '../renderers/shaders/UniformsUtils';
+import { ShaderLib } from '../renderers/shaders/ShaderLib';
+
+
+//custom depth and distance material to be attached to meshes
+
+var depthMaterialTemplate = new MeshDepthMaterial();
+
+depthMaterialTemplate.depthPacking = RGBADepthPacking;
+
+depthMaterialTemplate.clipping = true;
+
+depthMaterialTemplate.defines = {
+
+	INSTANCE_TRANSFORM: ''
+
+};
+
+var 
+	
+	distanceShader = ShaderLib[ "distanceRGBA" ],
+	distanceUniforms = UniformsUtils.clone( distanceShader.uniforms ),
+	distanceMaterialTemplate = new ShaderMaterial( {
+		defines: {
+			'USE_SHADOWMAP': '',
+			'INSTANCE_TRANSFORM': ''
+		},
+		uniforms: distanceUniforms,
+		vertexShader: distanceShader.vertexShader,
+		fragmentShader: distanceShader.fragmentShader,
+		clipping: true
+	})
+;
+
+//main class
+function InstancedMesh ( bufferGeometry , material , numInstances , dynamic , colors , uniformScale ) {
+
+	Mesh.call( this , (new InstancedBufferGeometry()).copy( bufferGeometry ) ); //hacky for now
+
+	this._dynamic = !!dynamic; //TODO: set a bit mask for different attributes?
+
+ 	this._uniformScale = !!uniformScale;
+
+ 	this._colors = !!colors;
+
+	this.numInstances = numInstances;
+
+	this._setAttributes();
+
+	/**
+	 * use the setter to decorate this material
+	 * this is in lieu of changing the renderer
+	 * WebGLRenderer injects stuff like this
+	 */
+	this.material = material.clone();
```

Comment:
this is all super questionable, in order to not have logic in the renderer that sets up the shaders properly, and flags on materials, this hack is happening
otherwise something like
```
myMaterialForInstancing = new THREE.SomeMaterial()
myMaterialForInstancing.instanced = true
```